src/rascal/models/sparse_kernels.hh

```diff
@@ -0,0 +1,498 @@
+        } else {
+          throw std::runtime_error("Given target_type " +
+                                   hypers["target_type"].get<std::string>() +
+                                   " is not known."
+                                   " It is either 'Structure' or 'Atom')");
+        }
+      } else {
+        throw std::runtime_error(
+            R"(No target_type given. It is either 'Structure' or Atom')");
+      }
+
+      auto kernel_type_str = hypers.at("name").get<std::string>();
+      if (kernel_type_str == "GAP") {
+        this->kernel_type = SparseKernelType::GAP;
+        this->kernel_impl =
+            make_sparse_kernel_impl<SparseKernelType::GAP>(hypers);
+      } else {
+        throw std::logic_error("Requested SparseKernel \'" + kernel_type_str +
+                               "\' has not been implemented.  Must be one of" +
+                               ": \'GAP\'.");
+      }
+    }
+
+    //! Move constructor
+    SparseKernel(SparseKernel && other) noexcept
+        : identifiers{std::move(other.identifiers)}, parameters{},
+          target_type{std::move(other.target_type)},
+          kernel_type{std::move(other.kernel_type)}, kernel_impl{std::move(
+                                                         other.kernel_impl)} {
+      this->parameters = std::move(other.parameters);
+    }
+
+    /*
+     * The root compute kernel function. It computes the kernel between 2 set of
+     * structures for a given representation specified by the calculator.
+     *
+     * @param calculator the calculator which has been used to calculate
+     * the representation on the two managers
+     * has been registered in the elements of managers and managers_b
+     * @param managers a ManagerCollection or similar collection of
+     * structure managers
+     * @param managers_b a ManagerCollection or similar collection of
+     * structure managers or a set of pseudo points
+     */
+    template <class Calculator, class StructureManagers, class SparsePoints>
+    math::Matrix_t compute(const Calculator & calculator,
+                           const StructureManagers & managers,
+                           const SparsePoints & sparse_points) {
+      using ManagerPtr_t = typename StructureManagers::value_type;
+      using Manager_t = typename ManagerPtr_t::element_type;
+      using Property_t = typename Calculator::template Property_t<Manager_t>;
+      auto && representation_name{calculator.get_name()};
+      using internal::TargetType;
+
+      switch (this->target_type) {
+      case TargetType::Structure:
+        return this->compute_helper<Property_t, TargetType::Structure>(
+            representation_name, managers, sparse_points);
+      case TargetType::Atom:
+        return this->compute_helper<Property_t, TargetType::Atom>(
+            representation_name, managers, sparse_points);
+      default:
+        throw std::logic_error("The combination of parameter is not handled.");
```

Comment:
This error message keeps on sneaking in, despite it being unreachable code. `TargetType` is an enum; it will only ever go to the cases enumerated in `TargetType`, all of which are covered in this switch statement.

Is the purpose of this message to catch programmer error, where we add a case to the enum but forget to add it to the switch statement? In which case, shouldn't the error message be more informative, giving the value of the enum that wasn't handled? And in any case, aren't there static analysis tools capable of finding such predictable mistakes?

Reply:
having a default in a switch over an `enum` disables the compiler warnings about missing enum values list of `case:`. Should I remove the default here ?

Reply:
I like the idea of having the logic error to tell you when you forgot to add it. I would leave it.

Reply:
made modification of the text